Smart Sampling for Monitoring

imandrealombardo opened the pull request:
### Description
Implemented a pipeline step to smart sample images from a database of client images. This is to re-train and evaluate the model.

### Related Jira Ticket
- Ticket: [BCV-52](https://gemeente-amsterdam.atlassian.net/browse/BCV-52)
- Description: Create script to perform the following features:
    - Randomly sample 10 images from each batch of images
    - Store the images in a folder 
    - Classify images based on detection counts and put in different bins
    - For each bin, randomly sample images for retraining
    - Store images for retraining in daytime folder which matches the starting time of the pre-inference pipeline

imandrealombardo committed:
Add filtering by conf_score

imandrealombardo committed:
trying to fix a bug with paths

imandrealombardo committed:
sample images for checking and retraining

imandrealombardo committed:
Count n° of detections per image

imandrealombardo committed:
group images in equal size bins

imandrealombardo committed:
smart sampling with bins

The code groups images in bins by number of detections, and then samples an equal amount of them from each bin. This is repeated for each date.

imandrealombardo committed:
refactoring of functions

Small refactoring of collect_images_above_threshold_from_db to enhance readability

imandrealombardo committed:
refactor smart_sampling()

imandrealombardo committed:
removed the hardcoded values

Now the important values (conf_score_threshold, sample_ratio, quality_check_sample_size) can be set from the config.yml

imandrealombardo committed:
logging test

imandrealombardo committed:
Merge branch 'main' into feature/BCV-52_Andrea

imandrealombardo committed:
Merge branch 'main' into feature/BCV-52_Andrea

imandrealombardo committed:
Improve readibility with docstrings and type hints

Also configure logging instead of using prints

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:
```diff
@@ -55,37 +76,337 @@ def smart_sampling(
+        The number of bins to be used for categorization.
+    """
+    
+    if detection_range <= 10:
```
And the number of bins?

imandrealombardo replied:
These numbers are dependent on the test images results: on 2000 images with a number of detections per image ranging from 0 to 242, 10 bins allowed for a good distribution of the images.

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:
```diff
@@ -55,37 +76,337 @@ def smart_sampling(
+
+def initialize_bin_counts(bins: np.ndarray) -> Tuple[Dict[str, List], List[str]]:
+    """
+    Initializes the counts and labels for bins.
```
Can you explain better what this function does?

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:
```diff
@@ -55,37 +76,337 @@ def smart_sampling(
+    bins : np.ndarray
+        The array of bin edges.
+    bin_labels : List[str]
+        A list of bin labels.
```
What are the bin labels? Can you make me an example?

imandrealombardo replied:
Each bin label is a `string` descriptor of the range of detection counts that belongs to each respective bin. When images are processed and their detection counts are categorized, they are placed into these bins based on where their count falls in these ranges. 

For example, if we have a range of detection counts from 0 to 100, we decide to create bins like [0-9], [10-19], ...., [90-99], [100]. The text representing each range (like "0-9" or "10-19") is a bin label. In this example, an image with 15 detections would fall into the bin labeled "10-19".

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:
```diff
@@ -55,37 +76,337 @@ def smart_sampling(
+        # Copy the sampled images
+        for image in images:
+            formatted_upload_date = upload_date.strftime("%Y-%m-%d_%H_%M_%S")
+            image_filename = image[2]  # Assuming image is a tuple (customer_name, upload_date, image_name)
```
All these assumptions of this data type you're using make the code a bit confusing for a person who tries to read it for the first time without any knowledge.
I would suggest to start creating your own data type to represent an Image, and then we can see how to simplify everything a bit further.

sdavrieux replied:
This data object could contain the details of the image you're using and for example also the number of detections and stuff like that.

epureanudiana replied:
I agree! There are a lot of functions and it's difficult to see how they all come together :)

Firstly, I suggest you move all of these functions outside of the `cleaning_pipeline/components/smart_sampling.py` and put them under `cleaning_pipeline/source/sampling_pipeline.py`. We usually have the `components/*.py `to only contain minimal code.

This way the components part remains clean, something like this: 

 ```
 @command_component(name="smart_sampling")
 def smart_sampling(
     input_structured_folder: Input(type=AssetTypes.URI_FOLDER),  # type: ignore # noqa: F821
     customer_cvt_folder: Output(type=AssetTypes.URI_FOLDER),  # type: ignore # noqa: F821
     customer_quality_check_folder: Output(type=AssetTypes.URI_FOLDER),  # type: ignore # noqa: F821
     customer_retraining_folder: Output(type=AssetTypes.URI_FOLDER),  # type: ignore # noqa: F821
     database_parameters_json: str,
     customer_name: str,
 ):
 
 SmartSampling.sample_images_for_quality_check(...)
 SmartSampling.sample_images_for_retraining(...)
 
 ```

You can create a class for sampling; in the `__init__` you can set attributes that you are now loosely collecting, for example `sample_images_for_retraining`, `n_images_to_sample` and you can also call the rest of the functions in there, whatever you prefer.

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:
```diff
@@ -104,54 +425,140 @@ def group_files_by_date(strings):
-    db_username = database_parameters["db_username"]
-    db_name = database_parameters["db_name"]
-    db_hostname = database_parameters["db_hostname"]
+    db_username = database_parameters.get("db_username")
```
Why get instead of []?

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:
```diff
@@ -104,54 +425,140 @@ def group_files_by_date(strings):
 
     return random_result
 
+def connect_to_database(database_parameters_json):
```
What is the reason to split the connect_to_database?

imandrealombardo replied:
The purpose was to make the database connection code reusable throughout the code instead of repeating it and to enhance readibility. In this case we only connect to the database once, so it may not be needed after all

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:
```diff
@@ -104,54 +425,140 @@ def group_files_by_date(strings):
+    except Exception as e:
+        logger.error(f"An unexpected error occurred: {e}")
+
+    return images_statistics, image_counts
```
The image counts can be counted from image statistics no?
What if instead of using all these multiples arrays with information that makes the code hard to follow we just use a dataframe with the actual data and we just share that across the functions?

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/submit_cleaning_pipeline.py:
```diff
@@ -64,6 +76,7 @@ def cleaning_pipeline():
     # Retrieve values from the YAML
     BlurringAsAServiceSettings.set_from_yaml("config.yml")
     settings = BlurringAsAServiceSettings.get_settings()
+    sampling_settings = settings["sampling_parameters"]
```
Where is this used?

sdavrieux commented on blurring_as_a_service/settings/settings_schema.py:
```diff
@@ -111,6 +111,12 @@ class InferenceCustomerPipelineSpec(SettingsSpecModel):
     model_parameters: InferenceModelParameters
+    
+
+class SmartSamplingPipelineSpec(SettingsSpecModel):
```
Good job :wink:

sdavrieux commented on config.example.yml:
```diff
@@ -85,4 +85,4 @@ logging:
     datefmt: "%Y-%m-%d %H:%M:%S"
-  ai_instrumentation_key: "APPLICATION_INSIGHTS_INSTRUMENTATION_KEY"
\ No newline at end of file
+  ai_instrumentation_key: "APPLICATION_INSIGHTS_CONNECTION_STRING"
```
Missing end of the file newline.

epureanudiana replied:
`"smart_sampling"` should be added to the list: 

```
own_packages: ["__main__", "blurring_as_a_service", "yolov5", "evaluate_with_cvt_metrics", "validate_model",
                 "evaluate_with_coco", "detect_and_blur_sensitive_data", "move_files", "split_workload"]
```
                 
did it work without it when you tested it?

imandrealombardo committed:
Update config.example.yml

imandrealombardo committed:
Update config.example.yml

epureanudiana commented on blurring_as_a_service/utils/generics.py:
```diff
@@ -40,8 +40,10 @@ def delete_file(file_path):
-    destination_path = os.path.join(output_path, relative_path)
+
+    # We don't use os.path.join because it doesn't work with Azure paths
+    source_path = input_path + relative_path
```
how did it work before then?

imandrealombardo replied:
I don't know, but I was stuck for days because using `os.path.join()` didn't create the correct path (it was ignoring `input_path` and `output_path`). The only way I found to make it work was to concatenate the two paths with a `+`!

imandrealombardo committed:
Replace .get with []

imandrealombardo committed:
Explain how initialize_bin_counts() works

imandrealombardo committed:
remove unused variable in submit_cleaning_pipeline

imandrealombardo committed:
Create SmartSampling class

imandrealombardo committed:
fix mistake and add functions in SmartSampling

imandrealombardo committed:
complete abstraction of functions from component

imandrealombardo committed:
fix mistake

imandrealombardo committed:
use DataFrame instead of multiple arrays

Still missing to fix the error of the database pulling all the images instead of filtering by grouped_images_by_date

imandrealombardo committed:
Finish creating class and using DataFrame

imandrealombardo committed:
docstrings

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:
```diff
@@ -96,300 +87,21 @@ def smart_sampling(
-        database_parameters_json, grouped_images_by_date, customer_name, conf_score_threshold
-    )
+    # Define the SmartSampling object
+    smartSampling = SmartSampling(input_structured_folder, customer_quality_check_folder, customer_retraining_folder, database_parameters_json,
```
We use underscores and not camelCase 😉

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py:
```diff
@@ -0,0 +1,458 @@
+
+logger = logging.getLogger(__name__)
+
+class SmartSampling:
```
This is a SmartSampler not the SmartSampling

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py:
```diff
@@ -0,0 +1,458 @@
+        A dictionary containing parameters for the sampling process, such as sample size, confidence score 
+        threshold, and sampling ratio.
+
+    Methods
```
We can remove the methods from here. 
Examples: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html

epureanudiana commented on blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py:
```diff
@@ -3,36 +3,47 @@
-    find_image_paths,
+from blurring_as_a_service.cleaning_pipeline.source.sampling_pipeline import (  # noqa: E402
+    SmartSampling,
+)
```
this import should come after line 39 where we setup the logger

imandrealombardo committed:
pre-hook fix

imandrealombardo committed:
move SmartSampling import

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py:
```diff
@@ -0,0 +1,458 @@
+        self.customer_name = customer_name
+        self.sampling_parameters = sampling_parameters
+        
+    def sample_images_for_quality_check(self, grouped_images_by_date: Dict[str, List[str]] 
```
All methods that are "private" should be underscored.

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py:
```diff
@@ -0,0 +1,458 @@
+        
+        logger.info(f'Sampling {n_images_to_sample} images for quality check.. \n')
+        
+        quality_check_images = SmartSampling.get_n_random_images_per_date(grouped_images_by_date, n_images_to_sample)
```
If a method is static you need to add the decorator and remove the self.
If the method is not static then it should be called by the same object, so self.method().
Otherwise you are calling it as a class method.

imandrealombardo committed:
putting collect_images function as not static

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py:
```diff
@@ -0,0 +1,458 @@
+        None
+        """
+        
+        # Collect images above the confidence score threshold from the database
```
Comment is redundant due to good naming on function and variables.
I'm a big fan (even if some people don't agree) of a no comments between the code, my theory is that if you need to write a comment then it means that that code needs to be refactored on a better structure.

imandrealombardo committed:
Remove methods from docstring

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py:
```diff
@@ -0,0 +1,458 @@
+                                images_statistics[upload_date] = extracted_data
+                                
+                # Flatten the dictionary to a list of dictionaries
+                flat_list = [item for sublist in images_statistics.values() for item in sublist]
```
This is an example of something that needs a comment and therefore should be refactored into a function for clarity. Possibly in a utils file where if someone needs to redo this in the future can just call the same function.

imandrealombardo committed:
use _ instead of camelCase

imandrealombardo committed:
change name from SmartSampling to SmartSampler

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py:
```diff
@@ -0,0 +1,458 @@
+            # Check if 'bin_label' column exists
+            if 'bin_label' not in df_date.columns:
+                logger.error("bin_label column not found in DataFrame for date: " + str(upload_date))
+                continue  # Skip to next date or handle error as needed
```
Not a big fan of continues and breaks, there is a lot of discussion if they are bad practice or not. 
I found easier to follow the flow of the program avoiding them.
@epureanudiana what is your idea about it?

epureanudiana replied:
yea, previously we replaced it with if else

sdavrieux commented on blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py:
```diff
@@ -0,0 +1,458 @@
+        
+        if detection_range <= 10:
+            return 3
+        elif 10 < detection_range <= 50:
```
else is irrelevant since there is a return, not an error but just a style thing :P
Also you could do:
```
if detection_range > 50:
   return 10
if detection_range > 10:
   return 5
return 3
```

imandrealombardo committed:
make bin labels line as a function

imandrealombardo committed:
style correction for if else

imandrealombardo committed:
Remove redundant comments

imandrealombardo committed:
put flatten dict method in utils

imandrealombardo committed:
change visibility of methods in SmartSampler

imandrealombardo committed:
encapsulate sampling of images in a function

imandrealombardo commented on blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py:
```diff
@@ -0,0 +1,458 @@
+                
+                # Adjust the number of images to sample from this bin
+                num_to_sample = min(images_per_bin + (1 if remainder > 0 else 0), len(df_bin))
+                remainder -= 1 if remainder > 0 else 0
```
I made a function out of this to explain it

imandrealombardo committed:
fix docstrings

imandrealombardo committed:
fix mistake

epureanudiana commented on blurring_as_a_service/cleaning_pipeline/submit_cleaning_pipeline.py:
```diff
@@ -33,13 +33,25 @@ def cleaning_pipeline():
     )
-    smart_sampling_step.outputs.customer_cvt_folder = Output(
+
+    smart_sampling_step.outputs.customer_retraining_folder = Output(
```
can you please add a description for the Inputs and Outputs objects? also the ones that are commented from the deletion phase,

like this:
```
 smart_sampling_step.outputs.customer_quality_check_folder = Output(
         type=AssetTypes.URI_FOLDER,
         mode="rw_mount",
         path=customer_quality_check_path,
         description="a short description"
```

imandrealombardo committed:
add description of inputs and outputs

imandrealombardo committed:
replace continue with if else loop

sdavrieux reviewed (approved):
Awesome job! Congrats on your first PR!
